File: java/0003_longest_substring_without_repeating_characters.java
/*
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/submissions/1287298218/
 */

import java.util.Set;
import java.util.HashSet;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<> ();
        int max = 0;
        int j = 0;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            while (set.contains(c)) {
                set.remove(s.charAt(j++));
            }
            set.add(c);
            max = Math.max(max, set.size());
        }
        return max;
    }
}

File: python/0003_longest_substring_without_repeating_characters.py
'''
https://leetcode.com/problems/longest-substring-without-repeating-characters/submissions/1663144470/
'''

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        v = set()
        i = -1
        max_j = 0
        for j in range(len(s)):
            while s[j] in v:
                i += 1
                v.remove(s[i])
            v.add(s[j])
            max_j = max(max_j, j-i)
        return max_j